builtin/logical/totp/path_roles.go

```diff
@@ -0,0 +1,342 @@
+	return &framework.Path{
+		Pattern: "keys/" + framework.GenericNameRegex("name"),
+		Fields: map[string]*framework.FieldSchema{
+			"name": {
+				Type:        framework.TypeString,
+				Description: "Name of the role.",
+			},
+
+			"generate": {
+				Type:        framework.TypeBool,
+				Default:     false,
+				Description: "Determines if a key is generated by vault or another service.",
+			},
+
+			"key_size": {
+				Type:        framework.TypeInt,
+				Default:     20,
+				Description: "Determines the size in bytes of the generated key.",
+			},
+
+			"key": {
+				Type:        framework.TypeString,
+				Description: "The shared master key used to generate a TOTP token.",
+			},
+
+			"issuer": {
+				Type:        framework.TypeString,
+				Description: `The name of the key's issuing organization.`,
+			},
+
+			"account_name": {
+				Type:        framework.TypeString,
+				Description: `The name of the account associated with the key.`,
+			},
+
+			"period": {
+				Type:        framework.TypeInt,
+				Default:     30,
+				Description: `The length of time used to generate a counter for the TOTP token calculation.`,
+			},
+
+			"algorithm": {
+				Type:        framework.TypeString,
+				Default:     "SHA1",
+				Description: `The hashing algorithm used to generate the TOTP token.`,
+			},
+
+			"digits": {
+				Type:        framework.TypeInt,
+				Default:     6,
+				Description: `The number of digits in the generated TOTP token.`,
+			},
+
+			"skew": {
+				Type:        framework.TypeInt,
+				Default:     0,
+				Description: `The number of delay periods that are allowed when validating a TOTP token.`,
+			},
+
+			"qr_size": {
+				Type:        framework.TypeInt,
+				Default:     200,
+				Description: `The pixel size of the generated square QR code.`,
+			},
+		},
+
+		Callbacks: map[logical.Operation]framework.OperationFunc{
+			logical.ReadOperation:   b.pathRoleRead,
+			logical.UpdateOperation: b.pathRoleCreate,
+			logical.DeleteOperation: b.pathRoleDelete,
+		},
+
+		HelpSynopsis:    pathRoleHelpSyn,
+		HelpDescription: pathRoleHelpDesc,
+	}
+}
+
+func (b *backend) Role(s logical.Storage, n string) (*roleEntry, error) {
+	entry, err := s.Get("key/" + n)
+	if err != nil {
+		return nil, err
+	}
+	if entry == nil {
+		return nil, nil
+	}
+
+	var result roleEntry
+	if err := entry.DecodeJSON(&result); err != nil {
+		return nil, err
+	}
+
+	return &result, nil
+}
+
+func (b *backend) pathRoleDelete(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	err := req.Storage.Delete("key/" + data.Get("name").(string))
+	if err != nil {
+		return nil, err
+	}
+
+	return nil, nil
+}
+
+func (b *backend) pathRoleRead(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	role, err := b.Role(req.Storage, data.Get("name").(string))
+	if err != nil {
+		return nil, err
+	}
+	if role == nil {
+		return nil, nil
+	}
+
+	switch role.Generate {
```

Comment:
I don't really like how doing a read on the role returns the underlying key data if it's been generated. Usually Vault doesn't expose that, it only exposes, basically, metadata. More broadly speaking, TOTP implementations in services (e.g. Google or Dropbox) generally only give you this information once (when you turn on 2FA) and never again -- if you want to remove it you need to disable TOTP and then turn it on again to generate a new code.

I think the logic that creates the barcode and returns the URL and so needs to move to the create function and be returned on create only. Then you have a single path here in the read function that returns the values currently used in the `default` case.